### ISSUE TITLE
Fails with angular-animate

### ISSUE THREAD

**Issue by csvan** [mattlewis92/angular-bluebird-promises#6]
Using this package in conjunction with angular-animate (1.4.2) and ui-router fails with the following stacktrace. Not sure how to replicate it other than running a plain project unfortunately.

UPDATE 2: the following most likely answers this. Promise() and $q() (`qFactory` in the Angular sources) work completely differently.

http://stackoverflow.com/questions/9822561/why-is-this-in-an-anonymous-function-undefined-when-using-strict

UPDATE: The problem appears to be the AnimateRunner  in angular-animate. Its prototype has the following method, which it itself calls recursively:

```
getPromise: function() {
      if (!this.promise) {
        var self = this;
        this.promise = $q(function(resolve, reject) {
          self.done(function(status) {
            status === false ? reject() : resolve();
          });
        });
      }
      return this.promise;
    },
```

At one point in this recursion, the `this` inside the Promise constructor becomes undefined:

```
function Promise(resolver) {
    if (typeof resolver !== "function") {
        throw new TypeError("the promise constructor requires a resolver function\u000a\u000a    See http://goo.gl/EC22Yn\u000a");
    }
    if (this.constructor !== Promise) { // error thrown here
       //....
    }
    //.....
```

I admit that my JS fu falters quite a bit here, but the following might lend some clues:

http://stackoverflow.com/questions/4011793/this-is-undefined-in-javascript-class-methods

Stacktrace:
```
TypeError: Cannot read property 'constructor' of undefined
    at Promise (http://localhost:3001/node_modules/bluebird/js/browser/bluebird.js:2243:13)
    at Object.t.module.directive.factory.factory.factory.c.getPromise (http://localhost:3001/node_modules/angular-animate/angular-animate.min.js:15:249)
    at Object.t.module.directive.factory.factory.factory.c.then (http://localhost:3001/node_modules/angular-animate/angular-animate.min.js:15:355)
    at Object.enter (http://localhost:3001/node_modules/angular-ui-router/release/angular-ui-router.min.js:7:22100)
    at http://localhost:3001/node_modules/angular-ui-router/release/angular-ui-router.min.js:7:22696
    at publicLinkFn (http://localhost:3001/node_modules/angular/angular.js:7417:29)
    at k (http://localhost:3001/node_modules/angular-ui-router/release/angular-ui-router.min.js:7:22678)
    at k.compile (http://localhost:3001/node_modules/angular-ui-router/release/angular-ui-router.min.js:7:23000)
    at invokeLinkFn (http://localhost:3001/node_modules/angular/angular.js:8652:9)
    at nodeLinkFn (http://localhost:3001/node_modules/angular/angular.js:8152:11)
    at compositeLinkFn (http://localhost:3001/node_modules/angular/angular.js:7543:13)
    at compositeLinkFn (http://localhost:3001/node_modules/angular/angular.js:7547:13)
    at compositeLinkFn (http://localhost:3001/node_modules/angular/angular.js:7547:13)
    at publicLinkFn (http://localhost:3001/node_modules/angular/angular.js:7418:30)
    at http://localhost:3001/node_modules/angular/angular.js:1635:27
    at Scope.$get.Scope.$eval (http://localhost:3001/node_modules/angular/angular.js:15848:28)
```

**Comment by csvan** [comment]
Alright, here is a hint. I can trace at least one of the errors back to this:

```
$stateProvider.state('app.admin', {
        url: '/admin',
        views: {
            'content@': {
                template: '<admin team="team"></admin>',
                controller: function controller($scope, team) {
                    $scope.team = team;
                },
                resolve: {
                    team: function team($q, Team, User) {
                        // Resolve the currently active team
                        // FIXME: Consider this a hack - we are not yet sure how to handle multiple teams
                        var deferred = $q.defer();
                        FrwrdUser.teams({
                            id: User.getCachedCurrent().id
                        }, function (teams, response) {
                            if (teams && teams.length > 0) {
                                console.log('ADMIN: found team');
                                deferred.resolve(teams[0]);
                            } else {
                                console.log('ADMIN: failed to find team');
                                deferred.reject('Failed to find team');
                            }
                        }, function (error) {
                            console.log('ADMIN: Error while resolving team');
                            deferred.reject('Failed to resolve team');
                        });

                        return deferred.promise;
                    }
                }
            }
        }
```

The problem seems to appear when deferred is used within the `resolve` property of ui-router.

**Comment by mattlewis92** [comment]
This should be fixed now in 0.5.3, let me know if you're still having any problems :)

**Comment by csvan** [comment]
Sorry for the late feedback, this is working great now!
